File: ppmdemo/rust/src/shader.rs
use std::env;
use std::fmt::Write as FmtWrite;
use std::fs;
use std::io::BufWriter;
use std::io::Write as IoWrite;
use std::ops::AddAssign;
use std::ops::{Add, Div, Mul, Sub};

#[derive(Default, Clone, Copy)]
struct Vec4 {
    x: f64,
    y: f64,
    z: f64,
    w: f64,
}

impl Vec4 {
    #[inline]
    fn new(x: f64, y: f64, z: f64, w: f64) -> Self {
        Self { x, y, z, w }
    }

    #[inline]
    fn sin(self) -> Self {
        let Self { x, y, z, w } = self;
        Self {
            x: x.sin(),
            y: y.sin(),
            z: z.sin(),
            w: w.sin(),
        }
    }

    #[inline]
    fn exp(self) -> Self {
        let Self { x, y, z, w } = self;
        Self {
            x: x.exp(),
            y: y.exp(),
            z: z.exp(),
            w: w.exp(),
        }
    }

    #[inline]
    fn tanh(self) -> Self {
        let Self { x, y, z, w } = self;
        Self {
            x: x.tanh(),
            y: y.tanh(),
            z: z.tanh(),
            w: w.tanh(),
        }
    }
}

impl Add<f64> for Vec4 {
    type Output = Self;

    fn add(self, rhs: f64) -> Self::Output {
        let Self { x, y, z, w } = self;
        Self {
            x: x + rhs,
            y: y + rhs,
            z: z + rhs,
            w: w + rhs,
        }
    }
}

impl Add<Vec4> for Vec4 {
    type Output = Self;

    fn add(self, rhs: Vec4) -> Self::Output {
        let Self { x, y, z, w } = self;
        let Self {
            x: rx,
            y: ry,
            z: rz,
            w: rw,
        } = rhs;
        Self {
            x: x + rx,
            y: y + ry,
            z: z + rz,
            w: w + rw,
        }
    }
}

impl Mul<f64> for Vec4 {
    type Output = Self;

    fn mul(self, rhs: f64) -> Self::Output {
        let Self { x, y, z, w } = self;
        Self {
            x: x * rhs,
            y: y * rhs,
            z: z * rhs,
            w: w * rhs,
        }
    }
}

impl Mul<Vec4> for f64 {
    type Output = Vec4;

    fn mul(self, rhs: Vec4) -> Self::Output {
        rhs * self
    }
}

impl AddAssign<Vec4> for Vec4 {
    fn add_assign(&mut self, rhs: Vec4) {
        *self = *self + rhs;
    }
}

impl Sub<Vec4> for f64 {
    type Output = Vec4;

    fn sub(self, rhs: Vec4) -> Self::Output {
        let Vec4 { x, y, z, w } = rhs;
        Vec4 {
            x: self - x,
            y: self - y,
            z: self - z,
            w: self - w,
        }
    }
}

impl Div<Vec4> for Vec4 {
    type Output = Self;

    fn div(self, rhs: Vec4) -> Self::Output {
        let Self { x, y, z, w } = self;
        let Self {
            x: rx,
            y: ry,
            z: rz,
            w: rw,
        } = rhs;
        Self {
            x: x / rx,
            y: y / ry,
            z: z / rz,
            w: w / rw,
        }
    }
}

#[derive(Default, Clone, Copy)]
struct Vec2 {
    x: f64,
    y: f64,
}

impl Vec2 {
    fn new(x: f64, y: f64) -> Self {
        Self { x, y }
    }

    fn yx(self) -> Self {
        let Self { x, y } = self;
        Self { x: y, y: x }
    }

    fn xyyx(self) -> Vec4 {
        let Self { x, y } = self;
        Vec4 {
            x: x,
            y: y,
            z: y,
            w: x,
        }
    }

    fn dot(self, rhs: Self) -> f64 {
        let Self { x, y } = self;
        let Self { x: rx, y: ry } = rhs;
        x * rx + y * ry
    }

    fn cos(self) -> Self {
        let Self { x, y } = self;
        Self {
            x: x.cos(),
            y: y.cos(),
        }
    }
}

impl Mul<f64> for Vec2 {
    type Output = Self;

    fn mul(self, rhs: f64) -> Self::Output {
        let Self { x, y } = self;
        Self {
            x: x * rhs,
            y: y * rhs,
        }
    }
}

impl Add<f64> for Vec2 {
    type Output = Self;

    fn add(self, rhs: f64) -> Self::Output {
        let Self { x, y } = self;
        Self {
            x: x + rhs,
            y: y + rhs,
        }
    }
}

impl Div<f64> for Vec2 {
    type Output = Self;

    fn div(self, rhs: f64) -> Self::Output {
        let Self { x, y } = self;
        Self {
            x: x / rhs,
            y: y / rhs,
        }
    }
}

impl Mul<Vec2> for f64 {
    type Output = Vec2;

    fn mul(self, rhs: Vec2) -> Self::Output {
        rhs * self
    }
}

impl Sub<Vec2> for Vec2 {
    type Output = Self;

    fn sub(self, rhs: Vec2) -> Self::Output {
        let Self { x, y } = self;
        let Self { x: rx, y: ry } = rhs;
        Self {
            x: x - rx,
            y: y - ry,
        }
    }
}

impl Add<Vec2> for Vec2 {
    type Output = Self;

    fn add(self, rhs: Vec2) -> Self::Output {
        let Self { x, y } = self;
        let Self { x: rx, y: ry } = rhs;
        Self {
            x: x + rx,
            y: y + ry,
        }
    }
}

impl Mul<Vec2> for Vec2 {
    type Output = Self;

    fn mul(self, rhs: Vec2) -> Self::Output {
        let Self { x, y } = self;
        let Self { x: rx, y: ry } = rhs;
        Self {
            x: x * rx,
            y: y * ry,
        }
    }
}

impl AddAssign<f64> for Vec2 {
    fn add_assign(&mut self, rhs: f64) {
        *self = *self + rhs;
    }
}

impl AddAssign<Vec2> for Vec2 {
    fn add_assign(&mut self, rhs: Vec2) {
        *self = *self + rhs;
    }
}

pub fn generate() {
    let scale = [16, 9];
    let fps = 60;
    let cell = fps;
    let w = scale[0] * cell;
    let h = scale[1] * cell;
    let rgb_max = 0xFF;
    let r = Vec2::new(w as f64, h as f64);
    let root = env::var("CARGO_MANIFEST_DIR").unwrap();
    let output_dir = root + "/output/shader";
    if !fs::exists(&output_dir).unwrap() {
        fs::create_dir_all(&output_dir).unwrap();
    }
    let mut buf = String::new();
    let header = format!("P6\n{} {}\n{}\n", w, h, rgb_max);
    let mut handles = Vec::with_capacity(fps);
    for i in 0..fps {
        FmtWrite::write_fmt(&mut buf, format_args!("{}/frame-{:02}.ppm", output_dir, i)).unwrap();
        let header = header.clone();
        let output_path = buf.clone();
        let handle = std::thread::spawn(move || {
            let file = fs::File::create(&output_path).unwrap();
            let mut file = BufWriter::new(file);
            IoWrite::write_all(&mut file, header.as_bytes()).unwrap();
            let t = i as f64 / fps as f64;
            for y in 0..h {
                for x in 0..w {
                    let fc = Vec2::new(x as f64, y as f64);
                    let p = (fc * 2. - r) / r.y;
                    let l = Vec2::default() + (4. - 4. * (0.7 - p.dot(p)).abs());
                    let mut v = p * l;
                    let mut o = Vec4::default();
                    for i in 1..=8 {
                        let i = Vec2::new(0., i as f64);
                        v += (v.yx() * i.y + i + t).cos() / i.y + 0.7;
                        o += (v.xyyx().sin() + 1.) * (v.x - v.y).abs();
                    }
                    o = (5. * (l.x - 4. - p.y * Vec4::new(-1., 1., 2., 0.)).exp() / o).tanh();
                    let buf = [
                        (o.x * rgb_max as f64) as u8,
                        (o.y * rgb_max as f64) as u8,
                        (o.z * rgb_max as f64) as u8,
                    ];
                    IoWrite::write_all(&mut file, &buf).unwrap();
                }
            }
            println!("Generated {}", output_path);
        });
        handles.push(handle);
        buf.clear();
    }
    for handler in handles {
        handler.join().unwrap();
    }
}

/// # Single-Threaded Implement
pub fn st_impl() {
    let scale = [16, 9];
    let fps = 60;
    let cell = fps;
    let w = scale[0] * cell;
    let h = scale[1] * cell;
    let rgb_max = 0xFF;
    let r = Vec2::new(w as f64, h as f64);
    let root = env::var("CARGO_MANIFEST_DIR").unwrap();
    let output_dir = root + "/shader";
    if !fs::exists(&output_dir).unwrap() {
        fs::create_dir(&output_dir).unwrap();
    }
    let mut buf = String::new();
    let header = format!("P6\n{} {}\n{}\n", w, h, rgb_max);
    for i in 0..fps {
        FmtWrite::write_fmt(&mut buf, format_args!("{}/frame-{:02}.ppm", output_dir, i)).unwrap();
        let output_path = &buf;
        let file = fs::File::create(output_path).unwrap();
        let mut file = BufWriter::new(file);
        IoWrite::write_all(&mut file, header.as_bytes()).unwrap();
        let t = i as f64 / fps as f64;
        for y in 0..h {
            for x in 0..w {
                let fc = Vec2::new(x as f64, y as f64);
                let p = (fc * 2. - r) / r.y;
                let l = Vec2::default() + (4. - 4. * (0.7 - p.dot(p)).abs());
                let mut o = Vec4::default();
                let mut v = p * l;
                for i in 1..=8 {
                    let i = Vec2::new(0., i as f64);
                    v += (v.yx() * i.y + i + t).cos() / i.y + 0.7;
                    o += (v.xyyx().sin() + 1.) * (v.x - v.y).abs();
                }
                o = (5. * (l.x - 4. - p.y * Vec4::new(-1., 1., 2., 0.)).exp() / o).tanh();
                let buf = [
                    (o.x * rgb_max as f64) as u8,
                    (o.y * rgb_max as f64) as u8,
                    (o.z * rgb_max as f64) as u8,
                ];
                IoWrite::write_all(&mut file, &buf).unwrap();
            }
        }
        println!("Generated {}", output_path);
        buf.clear();
    }
}

File: ppmdemo/c/shader.cpp
#include <math.h>
#include <stdio.h>

struct vec4 {
    float x, y, z, w;
    // vec4() : x(1.), y(1.), z(1.), w(1.) {}
    vec4() : x(0.), y(0.), z(0.), w(0.) {}
    vec4(float x, float y, float z, float w) : x(x), y(y), z(z), w(w) {}
};

struct vec2 {
    float x, y;
    // vec2() : x(1.), y(1.) {}
    vec2() : x(0.), y(0.) {}
    vec2(float x, float y) : x(x), y(y) {}
    vec2 yx() const { return vec2(y, x); }
    vec4 xyyx() const { return vec4(x, y, y, x); }
};

// vector v & scalar s
vec2 operator*(const vec2& v, float s) { return vec2(v.x * s, v.y * s); }
vec2 operator+(const vec2& v, float s) { return vec2(v.x + s, v.y + s); }
vec2 operator*(float s, const vec2& v) { return v * s; }
vec2 operator-(const vec2& v1, const vec2& v2) { return vec2(v1.x - v2.x, v1.y - v2.y); }
vec2 operator+(const vec2& v1, const vec2& v2) { return vec2(v1.x + v2.x, v1.y + v2.y); }
vec2 operator*(const vec2& v1, const vec2& v2) { return vec2(v1.x * v2.x, v1.y * v2.y); }
vec2 operator/(const vec2& v, float s) { return vec2(v.x / s, v.y / s); }
float dot(const vec2& v1, const vec2& v2) { return v1.x * v2.x + v1.y * v2.y; }
// vec2 abs(const vec2 &v) { return vec2(fabsf(v.x), fabsf(v.y)); }
vec2& operator+=(vec2& v1, const vec2& v2) {
    v1 = v1 + v2;
    return v1;
}
vec2& operator+=(vec2& v, float s) {
    v = v + s;
    return v;
}
vec2 cos(const vec2& v) { return vec2(cosf(v.x), cosf(v.y)); }
vec4 sin(const vec4& v) { return vec4(sinf(v.x), sinf(v.y), sinf(v.z), sinf(v.w)); }
vec4 exp(const vec4& v) { return vec4(expf(v.x), expf(v.y), expf(v.z), expf(v.w)); }
vec4 tanh(const vec4& v) { return vec4(tanhf(v.x), tanhf(v.y), tanhf(v.z), tanhf(v.w)); }
vec4 operator+(const vec4& v, float s) { return vec4(v.x + s, v.y + s, v.z + s, v.w + s); }
vec4 operator*(const vec4& v, float s) { return vec4(v.x * s, v.y * s, v.z * s, v.w * s); }
vec4 operator*(float s, const vec4& v) { return v * s; }
vec4 operator+(const vec4& v1, const vec4& v2) { return vec4(v1.x + v2.x, v1.y + v2.y, v1.z + v2.z, v1.w + v2.w); }
vec4& operator+=(vec4& v1, const vec4& v2) {
    v1 = v1 + v2;
    return v1;
}
vec4 operator-(float s, const vec4& v) { return vec4(s - v.x, s - v.y, s - v.z, s - v.w); }
vec4 operator/(const vec4& v1, const vec4& v2) { return vec4(v1.x / v2.x, v1.y / v2.y, v1.z / v2.z, v1.w / v2.w); }

int main() {
    char buf[256];
    int fps = 60;
    int cell = 60;
    int rgb_max = 255;
    int w = 16 * cell;
    int h = 9 * cell;
    vec2 r(w, h);  // resolution of screen
    for (int i = 0; i < fps; i++) {
        snprintf(buf, sizeof(buf), "output_%02d.ppm", i);
        const char* output_path = buf;
        FILE* file = fopen(output_path, "wb");
        fprintf(file, "P6\n");           // format
        fprintf(file, "%d %d\n", w, h);  // width / height
        fprintf(file, "%d\n", rgb_max);  // the maximum value of the color component, usually 255 for P6
        float t = (float)i / fps;
        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                vec4 o;         // an output of a single pixel
                vec2 fc(x, y);  // fragment coordinate
                vec2 p = (fc * 2. - r) / r.y;
                vec2 l;
                l += 4. - 4. * abs(.7 - dot(p, p));
                vec2 v = p * l;
                for (vec2 i; i.y++ < 8.; o += (sin(v.xyyx()) + 1.) * abs(v.x - v.y)) {
                    v += cos(v.yx() * i.y + i + t) / i.y + .7;
                }
                o = tanh(5. * exp(l.x - 4. - p.y * vec4(-1, 1, 2, 0)) / o);
                fputc(o.x * rgb_max, file);
                fputc(o.y * rgb_max, file);
                fputc(o.z * rgb_max, file);
            }
        }
        fclose(file);
        printf("Generated %s\n", output_path);
    }
    return 0;
}
